pinot-controller/src/main/java/com/linkedin/pinot/controller/api/resources/PqlQueryResource.java

```diff
@@ -68,44 +88,47 @@ public String get(
-
-      // Validate data access
-      AccessControl accessControl = _accessControlFactory.create();
-      if (!accessControl.hasDataAccess(httpHeaders, tableName)) {
-        return QueryException.ACCESS_DENIED_ERROR.toString();
-      }
+  public String getQueryResponse(String pqlQuery, String traceEnabled, HttpHeaders httpHeaders) {
+    // Get resource table name.
+    BrokerRequest brokerRequest;
+    try {
+      brokerRequest = REQUEST_COMPILER.compileToBrokerRequest(pqlQuery);
+    } catch (Exception e) {
+      LOGGER.info("Caught exception while compiling PQL query: {}, {}", pqlQuery, e.getMessage());
+      return QueryException.getException(QueryException.PQL_PARSING_ERROR, e).toString();
+    }
+    String tableName = TableNameBuilder.extractRawTableName(brokerRequest.getQuerySource().getTableName());
 
-      // Get brokers for the resource table.
-      List<String> instanceIds = _pinotHelixResourceManager.getBrokerInstancesFor(tableName);
-      if (instanceIds.isEmpty()) {
-        return QueryException.BROKER_RESOURCE_MISSING_ERROR.toString();
-      }
+    // Validate data access
+    AccessControl accessControl = _accessControlFactory.create();
+    if (!accessControl.hasDataAccess(httpHeaders, tableName)) {
+      return QueryException.ACCESS_DENIED_ERROR.toString();
+    }
 
-      // Retain only online brokers.
-      instanceIds.retainAll(_pinotHelixResourceManager.getOnlineInstanceList());
-      if (instanceIds.isEmpty()) {
-        return QueryException.BROKER_INSTANCE_MISSING_ERROR.toString();
-      }
+    // Get brokers for the resource table.
+    List<String> instanceIds = _pinotHelixResourceManager.getBrokerInstancesFor(tableName);
+    if (instanceIds.isEmpty()) {
+      return QueryException.BROKER_RESOURCE_MISSING_ERROR.toString();
+    }
 
-      // Send query to a random broker.
-      String instanceId = instanceIds.get(RANDOM.nextInt(instanceIds.size()));
-      InstanceConfig instanceConfig = _pinotHelixResourceManager.getHelixInstanceConfig(instanceId);
-      String url = "http://" + instanceConfig.getHostName().split("_")[1] + ":" + instanceConfig.getPort() + "/query";
-      return sendPQLRaw(url, pqlQuery, traceEnabled);
-    } catch (Exception e) {
-      LOGGER.error("Caught exception while processing get request", e);
-      return QueryException.getException(QueryException.INTERNAL_ERROR, e).toString();
+    // Retain only online brokers.
+    instanceIds.retainAll(_pinotHelixResourceManager.getOnlineInstanceList());
+    if (instanceIds.isEmpty()) {
+      return QueryException.BROKER_INSTANCE_MISSING_ERROR.toString();
     }
+
+    // Send query to a random broker.
+    String instanceId = instanceIds.get(RANDOM.nextInt(instanceIds.size()));
+    InstanceConfig instanceConfig = _pinotHelixResourceManager.getHelixInstanceConfig(instanceId);
+    String url = "http://" + instanceConfig.getHostName().split("_")[1] + ":" + instanceConfig.getPort() + "/query";
```

Comment:
this might break if there is an _ in the hostName. Better to replace Broker_ with "" or get everything after first occurrence of _